src/coreclr/jit/gentree.cpp

```diff
@@ -30407,6 +30598,8 @@ GenTree* Compiler::gtFoldExprHWIntrinsic(GenTreeHWIntrinsic* tree)
                     uint32_t result = BitOperations::LeadingZeroCount(static_cast<uint64_t>(value));
 
                     cnsNode->AsIntConCommon()->SetIconValue(static_cast<int32_t>(result));
+                    cnsNode->gtType = retType;
```

Comment:
why is this change?

Reply:
Without it the type of the node is incorrect. Arm64 expects a `TYP_INT` result but takes a `TYP_LONG` input.

Reply:
So was it broken until now? wondering why any test didn't catch this earlier.

Reply:
It's not been showing up as broken because of how 32-bit results work on most platforms. That is, the 32-bit and 64-bit registers are the same and setting the lower 32-bits zeros the upper 32-bits. So even with the node mistyped, the codegen and consumption in most scenarios was correct.

This got caught because of the new assert I added that validated we didn't accidentally change the return type as part of adding the TYP_MASK folding (since that could take in SIMD and need a MASK result)

Reply:
So it was a bug in that the IR was technically wrong, but because of how implicit conversions and the general operation actually worked things generally just worked as expected.